File: Ring_Oscillator.m
% Written by Max Ruiz
classdef RingOscillator
    properties
        vth;
        vtl;
        vmax;
    end
    
    methods
        function obj = RingOscillator(Vth, Vtl, Vmax)
            obj.vth = Vth;
            obj.vtl = Vtl;
            obj.vmax = Vmax;
        end
        
        function F = findFrequency(obj, tau)
            t = obj.findPeriod(tau);
            F = 1/t; 
        end
        
        function T = findPeriod(obj, tau)
            rt = obj.riseTime(tau);
            ft = obj.fallTime(tau);           
            T = rt + ft;       
        end
        
        function ft = fallTime(obj, tau)
            ft = -1*tau*log(obj.vtl/obj.vth);
        end
        
        function rt = riseTime(obj, tau)
            rtLow = -1*tau*log(1 - obj.vtl / obj.vmax);
            rtHigh = -1*tau*log(1 - obj.vth / obj.vmax);
            rt = rtHigh - rtLow;   
        end
        
    end
    
end

File: Ring_Oscillator.cpp
#include <cmath>

class Ring_Oscillator
{
private:
    double vth;
    double vtl;
    double vmax;
    double fallTime(double tau);
    double riseTime(double tau);

public:
    Ring_Oscillator(double Vth, double Vtl, double Vmax);
    double findFrequency(double tau);
    double findPeriod(double tau);

};

Ring_Oscillator::Ring_Oscillator(double Vth, double Vtl, double Vmax)
{
    this->vth = Vth;
    this->vtl = Vtl;
    this->vmax = Vmax;
}

double Ring_Oscillator::fallTime(double tau)
{
    double ft = -1*tau*log(this->vtl / this->vth);
    return ft;

}

double Ring_Oscillator::riseTime(double tau)
{
    double rtLow = -1*tau*log(1 - this->vtl / this->vmax);
    double rtHigh = -1*tau*log(1 - this->vth / this->vmax);
    double rt = rtHigh - rtLow;
    return rt;
}

double Ring_Oscillator::findPeriod(double tau)
{
    double rt = this->riseTime(tau);
    double ft = this->fallTime(tau);
    double T = rt + ft;
    return T;
}

double Ring_Oscillator::findFrequency(double tau)
{
    double T = this->findPeriod(tau);
    double F = 1/T;
    return F;
}
